aya/src/programs/mod.rs

```diff
@@ -971,3 +972,81 @@ impl ProgramInfo {
         Ok(ProgramInfo(info))
     }
 }
+
+/// Programs is an Iterator over loaded eBPF programs.
+pub struct Programs {
+    current: u32,
+    done: bool,
+}
+
+impl Iterator for Programs {
+    type Item = Result<ProgramInfo, ProgramError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.done {
+            return None;
+        }
+        let current = self.current;
+
+        match bpf_prog_get_next_id(current) {
+            Ok(Some(next)) => {
+                self.current = next;
+                Some(
+                    bpf_prog_get_fd_by_id(next)
+                        .map_err(|io_error| ProgramError::SyscallError {
+                            call: "bpf_prog_get_fd_by_id".to_owned(),
+                            io_error,
+                        })
+                        .and_then(|fd| {
+                            bpf_prog_get_info_by_fd(fd)
+                                .map_err(|io_error| ProgramError::SyscallError {
+                                    call: "bpf_prog_get_info_by_fd".to_owned(),
+                                    io_error,
+                                })
+                                .map(ProgramInfo)
+                        }),
+                )
+            }
+            Ok(None) => None,
+            Err((_, io_error)) => {
+                // If getting the next program failed, we have to yield None in our next
+                // iteration to avoid an infinite loop.
+                self.done = true;
+                Some(Err(ProgramError::SyscallError {
+                    call: "bpf_prog_get_fd_by_id".to_owned(),
+                    io_error,
+                }))
+            }
+        }
+    }
+}
+
+/// Returns an iterator over all loaded bpf programs.
+///
+/// This differs from [`crate::Bpf::programs`] since it will return all programs
+/// listed on the host system and not only programs loaded by Aya.
+///
+/// # Example
+/// ```
+/// # use aya::programs::iter_programs;
+///
+/// for p in iter_programs() {
+///     match p {
+///         Ok(program) => println!("{}", String::from_utf8_lossy(program.name())),
+///         Err(e) => println!("Error iterating programs: {:?}", e),
+///     }
+/// }
+/// ```
+///
+/// # Errors
+///
+/// Returns [`ProgramError::SyscallError`] if any of the syscalls required to either get
+/// next program id, get the program fd, or the [`ProgramInfo`] fail. In cases where
+/// iteration can't be performed, for example the caller does not have the necessary privileges,
+/// a single item will be yielded containing the error that occurred.
+pub fn iter_programs() -> Programs {
```

Comment:
I think that `loaded_programs` would be a better name for this. I don't think we
need to prepend `iter_` in the same way we don't for `HashMap::keys` etc